### PR TITLE
Allow date format yyyy-mm-dd

### PR DESCRIPTION
**Pull request by wucas**
This is related to #3355 in gap. If there are more places to change I am happy to change them as well. It seems there was something wrong in line 348, I added some brackets to correct the if statement.

### CODE REVIEW

**Comment by wucas** [comment]
I have not tested my changes though because I don't know how to make packages and I don't have one I could tests it with. The code that checks whether a date is valid I pushed to gap as well and there I tested it.

**Comment by wucas** [comment]
This is also related to Issue #20

**Comment by alex-konovalov** [comment]
Thank you! I will deal with this after https://github.com/gap-system/gap/pull/3355/ will stabilise.

**Comment by wucas** [comment]
I added the check for dates in the mm/dd/yyyy format. Different to the checks in `ValidatePackageInfo` it will print the invalid date" error additionally. I don't see a nice way to avoid this but I also think it's not too bad to have both errors printed since it is still clear what the problem ist.

**Comment by fingolfin** [comment]
@wucas to make a package, you can use https://github.com/gap-packages/PackageManager :-)

**Comment by alex-konovalov** [comment]
Thanks @wucas - is this now ready to be reviewed in conjunction with https://github.com/gap-system/gap/pull/3355 ?

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 1]
```diff
@@ -336,23 +336,34 @@ UpdatePackageInfoFiles := function(pkgdir)
...
-            Print("  ERROR (", info.PackageName, "): the date ", date,
-                  "should be a string of the form `dd/mm/yyyy'\n" );
-            has_error := true;
-          else
-            date := List( SplitString( date, "/" ), Int);
-            if not date[1] in [1..31] and
-                   date[2] in [1..12] and
-                   date[3] > 1999 and # GAP 4 appeared in 1999
-                   date[1] > DaysInMonth( date[2], date[3] ) then
+        if IsBound(info.Date) then
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: The old code used this:
```suggestion
        if IsBound(infon.Date) then
```
Which is right? What's the difference between `info` and `infon`? Perhaps we should add some comments to this code *sigh*

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 2]
```diff
@@ -336,23 +336,34 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+          if IsString(infon.date) and Length(infon.Date) = 10 and infon.Date{[3,6]} = "//" and
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Perhaps you can keep the `date := infon.Date;` assignment, and use `date` in a few of these lines -- that would reduce the diff, allowing us to focus on the important bits during review

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 3]
```diff
@@ -336,23 +336,34 @@ UpdatePackageInfoFiles := function(pkgdir)
                 "\n  This is not allowed, so the info file will not be changed\n");
           has_error := true;
         fi;
-        if IsBound(infon.Date) then
-          date := infon.Date;
-          if not IsString(date) and Length(date) = 10 and date{[3,6]} = "//" and
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Huh -- am I confused, or is this old code simply wrong? Seems as if parenthesis are missing after `not` and before `then` ... ?

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 4]
```diff
@@ -336,23 +336,34 @@ UpdatePackageInfoFiles := function(pkgdir)
...
-                    " is invalid\n" );
+                    "should be a string of the form `yyyy-mm-dd` or `dd/mm/yyyy'\n" );
               has_error := true;
             fi;
+          elif IsString(infon.Date) and Length(infon.Date) = 10 and infon.Date{[5,8]} = "--" and
+              ForAll( infon.Date{ [1,2,3,4,6,7,9,10] }, IsDigitChar ) then
+            date := List( SplitString( infon.Date, "-" ), Int);
+            date := Permuted(date, (1,3)); #sort such that date=[dd,mm,yyyy]
+          else
+            Print("  ERROR (", info.PackageName, "): the date ", date,
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Indentation:
```suggestion
             Print("  ERROR (", info.PackageName, "): the date ", date,
```

Also, I think `date` is not even bound here with your current code; another reason to add `date := infon.Date` back here.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 5]
```diff
@@ -336,23 +336,34 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+                    "should be a string of the form `yyyy-mm-dd` or `dd/mm/yyyy'\n" );
+              has_error := true;
+          fi;
+          if IsBound(date) and (not
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Perhaps like this would be better, to avoid warning twice?
```suggestion
          if not has_error and IsBound(date) and (not
```
Then again, `has_error` might have been set earlier... so alternatively, you could `Unbind(date);` above, after the errors are reported.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 1 above)
Comment (fingolfin) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: (just to clarify: the "sigh" did not refer to this PR, but rather towards the fact that the existing code is mostly undocumented)

Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: as far as I understand `infon` means new info (thus the n). The new info gets checked and later replaces the old info stored in `info`.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 3 above)
Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Yes I'm pretty sure that is wrong.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 4 above)
Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Oh right the way I did it I have to use `info.Date` there. Or change it as you suggested above.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 2 above)
Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: I have no problem with changing it, although the reason why I use `date` and `infon.Date` is that `date` is always a list with which I then check whether the date is valid and `info.Date` is a string storing the new date information. I thought it might be easier to read the code if I separate these. Otherwise the type of `date` would change.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 5 above)
Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: I use the alternative method with Unbind now.

**Comment by wucas** [comment]
I use `if IsBound( infon.Date )` now but as mentioned above I'am not entirely sure that this is how I should do it.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 6]
```diff
@@ -336,23 +336,46 @@ UpdatePackageInfoFiles := function(pkgdir)
...
-            ForAll( date{ [1,2,4,5,7,8,9,10] }, IsDigitChar ) then
-            Print("  ERROR (", info.PackageName, "): the date ", date,
-                  "should be a string of the form `dd/mm/yyyy'\n" );
-            has_error := true;
-          else
-            date := List( SplitString( date, "/" ), Int);
-            if not date[1] in [1..31] and
-                   date[2] in [1..12] and
-                   date[3] > 1999 and # GAP 4 appeared in 1999
-                   date[1] > DaysInMonth( date[2], date[3] ) then
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Oh well, this check also was wrong, and missing parenthesis after the `not` :-(

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 7]
```diff
@@ -336,23 +336,46 @@ UpdatePackageInfoFiles := function(pkgdir)
...
-                    " is invalid\n" );
+        if IsBound( infon.Date ) then
+          if IsString( infon.date ) and Length( infon.Date ) = 10 
+            and infon.Date{ [3,6] } = "//" 
+            and ForAll( infon.Date{ [1,2,4,5,7,8,9,10] }, IsDigitChar ) then
+            date := List( SplitString( infon.Date, "/" ), Int ); # here date=[dd,mm,yyyy]
+            # the format dd/mm/yyyy is ambigous and can be confused with mm/dd/yyyy 
+            # if it is clear from the date that the format is mm/dd/yyyy 
+            # print a message
+            if date[2] in [13..31] and date[1] in [1..12] then
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: So  `13/13/2020` would be OK? :-). Also you drop the check for 1999 but did not adjust the message. But why not keep the check? I'd put it into a newer local helper function, say,
```gap
IsPlausibleDate := date -> date[1] in [1..31] and
	                   date[2] in [1..12] and
	                   date[3] > 1999 and # GAP 4 appeared in 1999
	                   date[1] > DaysInMonth( date[2], date[3] );
```
Then you can write later on  `if not (IsBound(date) and IsPlausibleDate(date)) then ...` which IMHO is much more readable.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 8]
```diff
@@ -336,23 +336,46 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+              Print("  ERROR (", info.PackageName, "): the date ", infon.Date,
+                    Concatenation("should be a string of the form `yyyy-mm-dd`",
+                     "or `dd/mm/yyyy' representing a date since 1999\n" );
               has_error := true;
+              Unbind( date ); 
+              # so that we can check wheter an error was found by checking if
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]:
```suggestion
              # so that we can check whether an error was found by checking if
```
It was not immediately clear for me that this comment refers to the line before. I usually put those comments *before* the code in question, but I know this is not universal. Still, I'd do that, but also rephrase, perhaps like this:
```gap
  # unbind date so that later we can check ...
```

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 9]
```diff
@@ -336,23 +336,46 @@ UpdatePackageInfoFiles := function(pkgdir)
...
             fi;
+          elif IsString( infon.Date ) and Length( infon.Date ) = 10 
+            and infon.Date{ [5,8] } = "--" 
+            and ForAll( infon.Date{ [1,2,3,4,6,7,9,10] }, IsDigitChar ) then
+            date := List( SplitString( infon.Date, "-" ), Int);
+            date := date{ [3,2,1] }; # sort such that date=[dd,mm,yyyy]
+          else
+            Print("  ERROR (", info.PackageName, "): the date ", infon.Date,
+                    Concatenation("should be a string of the form `yyyy-mm-dd`",
+                     "or `dd/mm/yyyy' representing a date since 1999\n" );
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: I would remove the "Print" here, and also the one further up, and adjust the condition below...

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 10]
```diff
@@ -336,23 +336,46 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+              has_error := true;
+              Unbind( date ); 
+              # so that we can check wheter an error was found by checking if
+              # date is bound
+          fi;
+          if IsBound( date ) and ( not
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]:
```suggestion
          if not IsBound( date ) or ( not
```
Or go with `not (IsBound(date) and IsPlausibleDate(date)) ` as suggested above...

This way, you can remove the duplicate error message above, and also (if you like) those additional `has_error := true;` assignments, as the one below will be triggered anyway.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 7 above)
Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: At this point actually `13/13/2020` would be okay as this would be caught a couple of lines below when the validity of the date is actually checked. This check is only supposed to catch a date where it is plausible that the days and months are switched. In `13/13/2020` its just no date but e.g. in `12/24/2020` it is likely that it is a valid date but that the format is `mm/dd/yyyy`. I should make the comment above clearer though and also it makes no sense to print the same message twice for `12/24/2020`. But I like the idea that the whole check is smart enough to tell the user in such a situation that they've probably just switched days and months.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g (hunk 10 above)
Comment (wucas) [reply DistributionUpdate/PackageUpdate/PackageInfoTools.g]: If I found an error earlier I unbind `date`. Thus I should only check the validity of the date if `date` is bound. I'm not sure now why I want `not` in front of `IsBound( date )`. Does GAP lazy evaluation of these statements? Then I'd want it to stop checking the date as soon as it knows that `date` is not bound. But with your suggestion if `date` is unbound I'd still print the message and also try to access data that's not there.

**Comment by wucas** [comment]
I've made a few more changes based on @fingolfin suggestions. The idea is the following:
If no date is bound an according message is printed. Otherwise I check whether it is a string of the correct length, containing digits and `--` or `//` at the correct positions. If this is not the case a message is printed telling the user that `Date` should be a string of the format `dd/mm/yyyy` or `yyyy-mm-dd` representing a date since 1999. 
If this was the case then I split the string at `//` or `--` accordingly to get a list with which I can easily check the validity of the date. 
In the case of `//` I check whether it is plausible from the date that the format is `mm/dd/yyyy` and if so print a message telling the user specifically that they might switched month and day and that the format should be `dd/mm/yyyy` or `yyyy-mm-dd`. 
If either of this messages is printed no other message should be printed thus I set `has_error := true` and do `Unbind( date )`. 
If everything went well so far I check if the date is valid with a helper function for better readability. I the date is invalid another different message is printed.

tldr; My goal was to give the user an as specific as possible hint to what they might have done wrong. Thus four times `Print` now all with different messages.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 11]
```diff
@@ -336,23 +336,51 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+
+        # function to check wheter a date given as a list of three integers
+        # [day, month, year] is a plausible date of a release of a GAP package
+        IsPlausibleDate := date -> date[1] in [1..31] and
+	                   date[2] in [1..12] and
+	                   date[3] >= 1999 and # GAP 4 appeared in 1999
+	                   date[1] in [ 1..DaysInMonth( date[2], date[3] ) ];
+
+        if IsBound( infon.Date ) then
+          if IsString( infon.Date ) and Length( infon.Date ) = 10
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: minor optional suggestion:
```suggestion
          # look for date in format mm/dd/yyyy
          if IsString( infon.Date ) and Length( infon.Date ) = 10 
```

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 12]
```diff
@@ -336,23 +336,51 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+              Print("  ERROR (", info.PackageName, "): it seems that in the date ", infon.Date,
+                    Concatenation(" day and month are switched. ",
+                    "The date should be of the form `yyyy-mm-dd`",
+                    " or `dd/mm/yyyy`\n" ) );
+                    has_error := true;
+                    # unbind date so that we can check later whether an error 
+                    # was found by checking if date is bound
+                    Unbind( date );
+            fi;
+          elif IsString( infon.Date ) and Length( infon.Date ) = 10
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]:
```suggestion
          # look for date in ISO-8601 format yyyy-mm-dd
          elif IsString( infon.Date ) and Length( infon.Date ) = 10 
```

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 13]
```diff
@@ -336,23 +336,51 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+            Print("  ERROR (", info.PackageName, "): the date ", infon.Date,
+                    Concatenation(" should be a string of the form `yyyy-mm-dd`",
+                     " or `dd/mm/yyyy' representing a date since 1999\n" ) );
               has_error := true;
-            fi;
+              # unbind date so that we can check later whether an error 
+              # was found by checking if date is bound
+              Unbind( date ); 
+          fi;
+          if IsBound( date ) and ( not IsPlausibleDate( date ) ) then
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]:
```suggestion
          if IsBound( date ) and not IsPlausibleDate( date ) then
```

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 14]
```diff
@@ -336,23 +336,51 @@ UpdatePackageInfoFiles := function(pkgdir)
...
+                Print("  ERROR (", info.PackageName, "): the date ", infon.Date,
+                      Concatenation( " seems not to be a valid date since ","1999\n" ) );
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: Huh? Something isn't quite right here.

File: DistributionUpdate/PackageUpdate/PackageInfoTools.g [hunk 15]
```diff
@@ -336,23 +336,51 @@ UpdatePackageInfoFiles := function(pkgdir)
...
-            date := List( SplitString( date, "/" ), Int);
-            if not date[1] in [1..31] and
-                   date[2] in [1..12] and
-                   date[3] > 1999 and # GAP 4 appeared in 1999
-                   date[1] > DaysInMonth( date[2], date[3] ) then
-              Print("  ERROR (", info.PackageName, "): the date ", date,
-                    " is invalid\n" );
+            Print("  ERROR (", info.PackageName, "): the date ", infon.Date,
+                    Concatenation(" should be a string of the form `yyyy-mm-dd`",
+                     " or `dd/mm/yyyy' representing a date since 1999\n" ) );
```
Comment (fingolfin) [inline DistributionUpdate/PackageUpdate/PackageInfoTools.g]: BTW, there is no need to use `Concatenation` in `Print`, as `Print` is variadic (and you already make use of it). Same applies for other `Print` invocations

**Comment by fingolfin** [comment]
I am completely happy with this now, thank you @wucas . So now it's up to @alex-konovalov to decide whether and when to merge this :-)

**Comment by alex-konovalov** [comment]
For the update, https://github.com/gap-system/gap/pull/3355 is now merged.
